### PR TITLE
Fix Ninja warning for PRECOMPILED_HEADERS [RFC]

### PR DESCRIPTION
**Pull request by class4kayaker**
Simple modification for removing warning with Ninja and precompiled headers noted in #1491.

### CODE REVIEW

**Comment by gassmoeller** [comment]
@bangerth can you take a look whether this also fixes your warning? Also, is this actually fixing the issue, or is it just disabling the warning, and at some point cmake will remove the deprecated functionality?

File: CMakeLists.txt [hunk 1]
```diff
@@ -302,6 +302,13 @@ ENDIF()
 
 IF (ASPECT_PRECOMPILE_HEADERS)
   MESSAGE(STATUS "Enabling cotire to precompile external header files.")
+
+  # Request new BYPRODUCTS behavior
+  CMAKE_MINIMUM_REQUIRED(VERSION 3.2.0)
```
Comment (bangerth) [inline CMakeLists.txt]: We can't do this -- deal.II still only requires cmake 2.8.12

Comment (bangerth) [inline CMakeLists.txt]: But we can do something like this (copied from elsewhere):
```
#
# We support all policy changes up to version 3.1.0. Thus, explicitly set
# all policies CMP0001 - CMP0054 to new for version 3.1 (and later) to
# avoid some unnecessary warnings.
#
IF( "${CMAKE_VERSION}" VERSION_EQUAL "3.1" OR
    "${CMAKE_VERSION}" VERSION_GREATER "3.1" )
  CMAKE_POLICY(VERSION 3.1.0)
ENDIF()
```

Comment (class4kayaker) [inline CMakeLists.txt]: I checked this and I was going around this in what was likely the incorrect manner.

**Comment by class4kayaker** [comment]
I found the reference to the fix at setting the policy under an issue on the cotire repo, and looked it up at [policy documentation](https://cmake.org/cmake/help/v3.3/policy/CMP0058.html). My thought had been to check that the new BYPRODUCTS was available. I looked, and I think the warning was appearing because cotire uses add_custom_command, which triggers the warning if the policy is not specified. My impression is that for the new behavior, no phony rules will be generated, so if a later inclusion does not yet support the use of BYPRODUCTS, the Ninja generator may complain on a fresh build due to the policy disabling the earlier workaround.

I looked, and ASPECT is depending on using old behavior in at least one location (CMP0037), so just doing a version check to set to a new CMake version requirement should probably not be done.

**Comment by tjhei** [comment]
What is the reason for setting all policies instead of just setting CMP0058? Finally, does setting this policy just remove the warning (and it might still break with ninja) or does it check that you don't violate the fact that you have to declare byproducts?

**Comment by bangerth** [comment]
I had looked it up earlier and setting the policy makes cmake create different code for ninja (no other generator is affected). The problem is of course that the backward compatibility in cmake may one day go away.

**Comment by class4kayaker** [comment]
In this case, we are specifying use of the new behavior, so the concern should only be for older CMake versions, which I cannot easily test.

The test command by contrast does specify the use of old behavior (CMP0037), and might be worth looking into alternate approaches.

**Comment by bangerth** [comment]
So @tjhei just talked about this and we think that what should actually happen is that the policy is set to `OLD`, not `NEW`. The `NEW` policy requires one to specify the byproducts explicitly. We don't do this correctly in the `tests/` subdirectory. We don't know (and have no way of ascertaining) whether cotire is doing this. The `OLD` policy, in contrast, just does a compatibility workaround. This is what we probably want, even if that's inefficient with ninja. Setting no policy at all uses the `OLD` policy, but warns about it.

**Comment by class4kayaker** [comment]
So if I understand correctly, you are suggesting that we explicitly go with the use of old (deprecated) behavior that may or may not still be maintained in the future. If this is the approach desired, I think leaving out the policy declaration would be the better approach, as it would keep the warning, but not

If cotire is at https://github.com/sakra/cotire, then the maintainer explicitly suggests this fix [here](https://github.com/sakra/cotire/issues/81#issuecomment-172300729), suggesting it does maintain the new behavior.

I will look into altering the test build commands to correctly use the new behavior.

**Comment by bangerth** [comment]
Hm, but doesn't setting the policy to `NEW` lead to wrong output for cases where rules don't properly describe `BYPRODUCT` files?

It may be that we never do that, and that cotire never does that. But then I'm surprised that we get the warning.

**Comment by class4kayaker** [comment]
I believe it is a case of if the behavior is not specified, a custom_command without a BYPRODUCTS clause will be raise the warning. Note that if the new behavior is specified, and a dependency is not given in an out of source build a warning will be generated (I found this by working on getting the tests to have properly specified byproducts).

**Comment by class4kayaker** [comment]
I have added the test products to the build using a conditional on the CMP0058 policy in the tests. The warning occurs when the add_custom_command or add_custom_target is not provided BYPRODUCTS after the policy is defined when an explicit behavior is not requested.

**Comment by gassmoeller** [comment]
Due to the sensitivity of this pull request, I would suggest waiting with merging it until after the hackathon. We do not want to cause new problems with the tester. Still @bangerth or @tjhei could you check if this solves the warning with ninja? Thanks @class4kayaker for improving this!

**Comment by class4kayaker** [comment]
This pull request appears to be based on aspects that have been adjusted since its creation, suggesting that I should look into . I will look into what may still be necessary.

The tests aspect of this pull request has been handled externally (#1573), so the only element that is active here is the use of CMP0058 in cotire

**Pull request closed without merging**